sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5247,7 +5247,8 @@ pi_result _pi_event::cleanup(pi_queue LockedQueue) {
     // NOTE: that this needs to be done only once for an event so
     // this is guarded with the CleanedUp flag.
     //
-    PI_CALL(EventRelease(this, LockedQueue));
+    if (RefCount > 0)
```

Comment:
Why is the event with "0" reference count is still flying around?

Reply:
We call cleanup() when the RefCount of an Event reaches zero.
```
 if (--(Event->RefCount) == 0) {
    if (!Event->CleanedUp)
      Event->cleanup(LockedQueue);
```

Reply:
Then do we call `cleanup()` when RefCount is > 0?

Reply:
I am just  answering to your question "Why is the event with "0" reference count is still flying around?".
We call cleanup() when RefCount becomes zero in piEventRelease(), and then We call EventRetain() in cleanup(), it is a circular dependency and causes a crash.
That's why we need to avoid calling piEventRelease() when cleanup() is called from piEventRelease().

Reply:
If `cleanup` is only called on RefCount == 0 then why do we call `EventRelease` from it? I am not understanding the _reason_ for the circular dependency now.

Reply:
No, cleanup() is called even when RefCount is not zero.
That's why there is a call to piEventRelease() from cleanup().
For example, cleanup() is called from piEventsWait() and pi_queue::resetCommandList().
If cleanup() has been called only when RefCount is zero, this code should have been already broken.

Reply:
OK, so `event.cleanup()` is called as long as the event is known to get completed. But the at cleanup() it will set the `CleanedUp` to true, so there should be no other `cleanup()` called on it (at destruction). There is a comment about it:

https://github.com/intel/llvm/blob/b01d820a5fc4d0df63fda946a9ee522f7f5dac9f/sycl/plugins/level_zero/pi_level_zero.cpp#L5330

Please explain why this check for `RefCount == 0` is necessary (show the exact scenario).

Reply:
@bso-intel : please follow up. I really don't want unnecessary RefCount checks which are indicative of some design misses if they are really needed.

Reply:
Oops, I forgot to remove that.
Done now.

Reply:
I think we should remove call to the cleanup from piEventRelease.
Event is retained explicitly when created in createEventAndAssociateQueue() to ensure that the event is not destroyed before it is really signaled. Ref count is decremented in Event->cleanup() after event is singaled. This
means that if reference count for an event is zero then cleanup was already called for this event and should not be called anymore. Interop events and user events don't have associated command list, queue,
kernel and waitlist, so we don't need to call cleanup() for interop and user
events.
I had an intention to do this earlier in https://github.com/intel/llvm/pull/5230/commits/94943f3493a9a09ce4cff9a486dc2b0c5429fedd but forgot about it.
Opened a PR for this small change: https://github.com/intel/llvm/pull/6073.